File: ports/elixir/lib/unicode_security/covert/surrogate_reassembly.ex
defmodule UnicodeSecurity.Covert.SurrogateReassembly do
  alias UnicodeSecurity.Utf8

  def looks_like_byte_stream?(input), do: Enum.all?(input, fn cp -> cp < 0x100 end)

  def detect(input) do
    bytes =
      input |> Enum.map(fn cp -> if cp > 0xFF, do: 0xFF, else: cp end) |> :binary.list_to_bin()

    case Utf8.first_invalid_utf8_offset(bytes) do
      nil -> %{sub: nil, positions: []}
      {offset, kind} -> %{sub: sub_threat(kind), positions: [offset]}
    end
  end

  defp sub_threat(:overlong_encoding), do: "Overlong"
  defp sub_threat(:surrogate_codepoint), do: "Cesu8"
  defp sub_threat(:truncated_sequence), do: "Truncated"
  defp sub_threat(:invalid_start_byte), do: "InvalidStartByte"
  defp sub_threat(:invalid_continuation_byte), do: "InvalidContinuation"
  defp sub_threat(:codepoint_beyond_max), do: "CodepointBeyondMax"
end

File: ports/ruby/lib/unicode_ruby/security/covert/surrogate_reassembly.rb
# frozen_string_literal: true

require_relative "../../utf8"
require_relative "../../strict"

module UnicodeRuby
  module Security
    module Covert
      # Surrogate-reassembly / malformed-byte-stream detection.  The input
      # codepoint list is treated as a byte stream (one octet per entry); the
      # family only applies when every entry is a byte (< 0x100).  The verdict
      # projects the first UTF-8 violation onto a covert-layer sub-threat.
      module SurrogateReassembly
        Detection = Struct.new(:sub, :positions)

        module_function

        # True iff every entry fits in one octet — the looksLikeByteStream gate.
        def looks_like_byte_stream(input)
          input.all? { |cp| cp < 0x100 }
        end

        def sub_threat_of_reject_kind(kind)
          case kind
          when Strict::OVERLONG_ENCODING then "Overlong"
          when Strict::SURROGATE_CODEPOINT then "Cesu8"
          when Strict::TRUNCATED_SEQUENCE then "Truncated"
          when Strict::INVALID_START_BYTE then "InvalidStartByte"
          when Strict::INVALID_CONTINUATION_BYTE then "InvalidContinuation"
          when Strict::CODEPOINT_BEYOND_MAX then "CodepointBeyondMax"
          else
            raise "surrogate_reassembly: unknown reject kind #{kind.inspect}"
          end
        end

        # Any value > 0xFF is clamped to 0xFF (never a valid UTF-8 start byte),
        # mirroring the Lean toBytes helper.
        def detect(input)
          bytes = input.map { |cp| cp > 0xFF ? 0xFF : cp }
          result = Utf8.first_invalid_offset(bytes)
          if result.nil?
            Detection.new(nil, [])
          else
            offset, kind = result
            Detection.new(sub_threat_of_reject_kind(kind), [offset])
          end
        end
      end
    end
  end
end
